### PR TITLE
[SPARK-22514][SQL] move ColumnVector.Array and ColumnarBatch.Row to individual files

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Logically the `Array` doesn't belong to `ColumnVector`, and `Row` doesn't belong to `ColumnarBatch`. e.g. `ColumnVector` needs to return `Array` for `getArray`, and `Row` for `getStruct`. `Array` and `Row` can return each other with the `getArray`/`getStruct` methods.

This is also a step to make `ColumnVector` public, it's cleaner to have `Array` and `Row` as top-level classes.

This PR is just code moving around, with 2 renaming: `ColumnVector.Array` -> `ColumnarArray`, `ColumnarBatch.Row` -> `ColumnarRow`.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @gatorsmile @kiszk

**Comment by kiszk** [comment]
I think that it is good to have top-level classes. I will check code later today.

I feel that  `VectorBased...` is not clear. IMHO, `Vector` and `Column` is orthogonal. `Vector` can be used with `Row`.
How about `ColumnBasedArray`/`ColumnBasedRow` or something? Other reviewers would find other names.

**Comment by viirya** [comment]
`ColumnVectorArray` and `ColumnVectorRow`?

**Comment by viirya** [comment]
Moving `ColumnVector.Array` and `ColumnarBatch.Row` to individual files is also good for reading column vector sources actually.

**Comment by cloud-fan** [comment]
`ColumnVectorArray` was my first idea but it sounds like array of column vector. If we don't mind to be verbose, how about `ArrayInColumnVector`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala [hunk 1]
```diff
@@ -179,7 +179,7 @@ class ColumnVectorSuite extends SparkFunSuite with BeforeAndAfterEach {
     testVector.putArray(2, 3, 0)
     testVector.putArray(3, 3, 3)
 
-    val array = new ColumnVector.Array(testVector)
+    val array = new VectorBasedArray(testVector)
 
     assert(array.get(0, arrayType).asInstanceOf[ArrayData].toIntArray() === Array(0))
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala]: nit: Can we use `getArray()` instead of `get()`?

**Comment by viirya** [comment]
`ColumnVectorArray` or `ArrayInColumnVector` both good to me.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/VectorBasedRow.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+import org.apache.spark.sql.catalyst.util.MapData;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.types.CalendarInterval;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Row abstraction in {@link ColumnVector}. The instance of this class is intended
+ * to be reused, callers should copy the data out if it needs to be stored.
+ */
+public final class VectorBasedRow extends InternalRow {
```
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/VectorBasedRow.java]: How about `ColumnarRow`?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/VectorBasedArray.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import org.apache.spark.sql.catalyst.util.MapData;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.types.CalendarInterval;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Array abstraction in {@link ColumnVector}. The instance of this class is intended
+ * to be reused, callers should copy the data out if it needs to be stored.
+ */
+public final class VectorBasedArray extends ArrayData {
```
Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/VectorBasedArray.java]: `ColumnarArray`?

**Comment by viirya** [comment]
LGTM

**Comment by kiszk** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**